language/move-lang/src/parser/ast.rs

```diff
@@ -0,0 +1,488 @@
+
+new_name!(ModuleName);
+
+#[derive(Debug, Hash, Eq, PartialEq, Ord, PartialOrd, Clone)]
+pub struct ModuleIdent_ {
+    pub name: ModuleName,
+    pub address: Address,
+}
+#[derive(Debug, Hash, Eq, PartialEq, Ord, PartialOrd, Clone)]
+pub struct ModuleIdent(pub Spanned<ModuleIdent_>);
+
+#[derive(Debug)]
+pub struct ModuleDefinition {
+    pub uses: Vec<(ModuleIdent, Option<ModuleName>)>,
+    pub name: ModuleName,
+    pub structs: Vec<(StructDefinition)>,
+    pub functions: Vec<(Function)>,
+}
+
+//**************************************************************************************************
+// Structs
+//**************************************************************************************************
+
+new_name!(Field);
+new_name!(StructName);
+
+pub type ResourceLoc = Option<Loc>;
+
+#[derive(Debug, PartialEq)]
+pub struct StructDefinition {
+    pub resource_opt: ResourceLoc,
+    pub name: StructName,
+    pub type_parameters: Vec<(Name, Kind)>,
+    pub fields: StructFields,
+}
+
+#[derive(Debug, PartialEq)]
+pub enum StructFields {
+    Defined(Vec<(Field, SingleType)>),
+    Native(Loc),
+}
+
+//**************************************************************************************************
+// Functions
+//**************************************************************************************************
+
+new_name!(FunctionName);
+
+#[derive(PartialEq, Debug)]
+pub struct FunctionSignature {
+    pub type_parameters: Vec<(Name, Kind)>,
+    pub parameters: Vec<(Var, SingleType)>,
+    pub return_type: Type,
+}
+
+#[derive(PartialEq, Debug, Clone)]
+pub enum FunctionVisibility {
+    Public(Loc),
+    Internal,
+}
+
+#[derive(PartialEq, Debug)]
+pub enum FunctionBody_ {
+    Defined(Sequence),
+    Native,
+}
+pub type FunctionBody = Spanned<FunctionBody_>;
+
+#[derive(PartialEq, Debug)]
+// (public?) foo<T1(: copyable?), ..., TN(: copyable?)>(x1: t1, ..., xn: tn): t1 * ... * tn {
+//    body
+//  }
+// (public?) native foo<T1(: copyable?), ..., TN(: copyable?)>(x1: t1, ..., xn: tn): t1 * ... * tn;
+pub struct Function {
+    pub visibility: FunctionVisibility,
+    pub signature: FunctionSignature,
+    pub acquires: Vec<SingleType>,
+    pub name: FunctionName,
+    pub body: FunctionBody,
+}
+
+//**************************************************************************************************
+// Types
+//**************************************************************************************************
+
+#[derive(Debug, PartialEq)]
+pub enum TypeName_ {
+    // N
+    Name(Name),
+    // M.S
+    ModuleType(ModuleName, StructName),
+    // OxADDR.M.S
+    QualifiedModuleType(ModuleIdent, StructName),
+}
+pub type TypeName = Spanned<TypeName_>;
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub enum Kind_ {
+    // Kind representing all types
+    Unknown,
+    // Linear resource types
+    Resource,
+    // Explicitly copyable types
+    Affine,
```

Comment:
As far as I understand it (could be wrong):

Linear types are types that must be used exactly once (this says nothing about copying)
  - Think of a Rust Result type. Must be used exactly once or you get a error (warning). But it can be cloned 
Affine types are types that can be used exactly once (this says nothing about copying)
  - Think of a Rust struct. Can be used once, but could be discarded. But it can also be cloned

Resources are linear types with this extra rule about copying

Reply:
Ah I get your point and they map directly to the various types of values in the language:
- Resources **must be used once** (linear) and cannot be copied.
- Structs are not by default copyable thus affine. They can be explicitly copied using a special syntax.
- Primitives like u64 can be used with almost no restrictions.

Thanks for the clarification. I'm fine with this.